modules/indexing/src/test/java/org/apache/ignite/internal/processors/cache/PrecisionTest.java

```diff
@@ -0,0 +1,344 @@
+            "   bin binary(5)" +
+            ") with \"CACHE_NAME=SQL,VALUE_TYPE=SQL_TYPE\""));
+
+        StringBuilder validStr = new StringBuilder("12345");
+
+        SqlFieldsQuery strQry = new SqlFieldsQuery("insert into TABLE(id, str) values (?, ?)");
+        SqlFieldsQuery binQry = new SqlFieldsQuery("insert into TABLE(id, bin) values (?, ?)");
+
+        try (IgniteClient client = Ignition.startClient(new ClientConfiguration().setAddresses("127.0.0.1"))) {
+            ClientCache cc = client.cache("SQL");
+
+            for (int i = 1; i < CNT; i++) {
+                validStr.append("0");
+
+                assertClientPrecision(new Callable<Object>() {
+                    @Override public Object call() {
+                        cc.query(strQry.setArgs(KEY, validStr.toString())).getAll();
+
+                        return null;
+                    } }, "STR", 5 + i);
+
+                assertClientPrecision(new Callable<Object>() {
+                    @Override public Object call() {
+                        cc.query(binQry.setArgs(KEY, validStr.toString().getBytes(StandardCharsets.UTF_8))).getAll();
+
+                        return null;
+                    } }, "BIN", 5 + i);
+
+                assertClientPrecision(new Callable<Object>() {
+                    @Override public Object call() {
+                        BinaryObject bo = ignite(0).binary().builder("SQL_TYPE")
+                            .setField("id", KEY)
+                            .setField("str", validStr.toString())
+                            .build();
+
+                        cc.put(0, bo);
+
+                        return null;
+                    } }, "STR", 5 + i);
+
+                assertClientPrecision(new Callable<Object>() {
+                    @Override public Object call() {
+                        BinaryObject bo = ignite(0).binary().builder("SQL_TYPE")
+                            .setField("id", KEY)
+                            .setField("bin", validStr.toString().getBytes(StandardCharsets.UTF_8))
+                            .build();
+
+                        cc.put(0, bo);
+
+                        return null;
+                    } }, "BIN", 5 + i);
+            }
+        }
+    }
+
+    /** */
+    @Test
+    public void testClientInsertValueVarColumns() throws Exception {
+        try (IgniteClient client = Ignition.startClient(new ClientConfiguration().setAddresses("127.0.0.1"))) {
+            ClientCache cc = client.cache("PERSON");
+
+            StringBuilder validStr = new StringBuilder("12345");
+
+            SqlFieldsQuery strQry = new SqlFieldsQuery("insert into PERSON(_KEY, name) values (?, ?)");
+            SqlFieldsQuery binQry = new SqlFieldsQuery("insert into PERSON(_KEY, arr) values (?, ?)");
+
+            for (int i = 1; i < CNT; i++) {
+                String v = validStr.append("0").toString();
+                int len = 5 + i;
+
+                assertClientPrecision(new Callable<Object>() {
```

Comment:
I am not insisting, but may be use here just lambda?

Reply:
Usage of pure lambda is broken for assertThrows. It converts lambda to RunnableX that rethrow any exception with IgniteException wrapper. So it's impossible to check exception class then.

Reply:
Works with lambda with return value. Thanks.